keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md

```diff
@@ -0,0 +1,1150 @@
+- Field selectors only exist for strings, enums, integers and booleans.
+- Field selectors for name and namespace are provided automatically.
+- Field selectors that do not match the path name of the actual field do exist,
+  for historical reasons, but the best practice is for field selectors to match
+  the exact path name of the field.
+
+### Plan
+
+Because `generic.StoreOptions` consumes selectors from built-in and CRDs in the
+same way, we would extend the `GetAttrs` of CRDs to offer fields for selection.
+https://github.com/jpbetz/kubernetes/tree/crd-object-filters contains a working
+prototype.
+
+We would leverage the fieldPath validation supported added for CRD Validation
+Rules to validate paths.
+
+We would leverage the jsonpath library in the same way as used by
+additionalPrinterColumns to retrieve values from custom resources using
+jsonpaths.
+
+The majority of the remaining work will be to update the API to add the
+field and to validate it.
+
+#### API
+
+```go
+type CustomResourceDefinitionVersion struct {
+  
+  // selectableFields specifies paths to fields that may be used as field selectors.
+  // A maximum of 8 selectable fields are allowed.
+	// See https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors
+	SelectableFields []SelectableField `json:"selectableFields,omitempty" protobuf:"bytes,9,rep,name=selectableFields"`
+}
+
+  // SelectableField specifies the JSON path of a field that may be used with field selectors.
+type SelectableField struct {
+	// jsonPath is a simple JSON path which is evaluated against each custom resource to produce a
+  // field selector value.
+  // Only JSON paths without the array notation are allowed.
+  // Must point to a valid field of type string, boolean or integer. Strings with enum values
+  // and strings with formats are allowed.
+  // Must not point to metdata fields.
+	JSONPath string `json:"jsonPath" protobuf:"bytes,1,opt,name=jsonPath"`
+
+  // valueIfAbsent provides a value for field selection if the value referred to
+  // by jsonPath is absent.
+  // valueIfAbsent is required unless all parts of jsonPath are required fields, or have defaults.
+  // +Optional
+  ValueIfAbsent *string `json:"valueIfAbsent" protobuf:"bytes,2,opt,name=valueIfAbsent"`
+}
+```
+
+##### Validaiton rules
+
+- `selectableFields[].jsonPath` must be a "simple path" (similar to
+  `additionalPrinterColumns[].jsonPath` except validated properly). I.e.
+    property names separated by '.', e.g. `.spec.x.y` (The path may not contain
+    map or array indexing)
+- The schema type referenced by `selectableFields[].jsonPath` must be one of:
+  string, integer, boolean (enums and fields with string formats are allowed)
+- If (and only if) the schema type referenced by `selectableFields[].jsonPath`
+  is optional (directly or indirectly), a `selectableFields[].ifAbsent` field
+  must provide a fallback string value.
+- A maximum of 8 `selectableFields` per CRD version are allowed.
```

Comment:
Why?  Seems like good guidance, but there's no clear performance related reason for this restriction since we iterate every item to decide, right?

Reply:
I'm uncertain what the max should be, but the watch cache indexes at least one special purpose field selector and  putting a limit offers us some headroom to perform more caching in the future.

One thing I'd like to avoid is having CRD authors write a script that adds a selectableField for every field in their CRD.. I'd like them be more selective in how they identify valuable access patterns.

I'm open to ideas of what a good limit might be.

cc @wojtek-t

Reply:
The watch cache index for pod.nodeName is used because it's a field selector that is extremely widely used (e.g. by kubelets). It's not a requirement - it's an optimization, so I don't think it should really guide us as a limiting factor.

I like the fact that it prevents users from blindly making every field selectable, but I'm not attached to any particular value of the limit.